apps/pipelines/nodes/nodes.py

```diff
@@ -72,31 +70,118 @@ class CreateReport(LLMResponse):
+        input = state["messages"][-1]
+        self.logger.debug(f"Returning input: '{input}' without modification")
+        return input
+
+
+class ExtractStructuredDataBasic(LLMResponse):
+    __human_name__ = "Extract structured data (Basic)"
+    data_schema: str
+
+    def _process(self, state: PipelineState) -> RunnableLambda:
+        json_schema = ExtractStructuredDataBasic.to_json_schema(json.loads(self.data_schema))
+        prompt = PromptTemplate.from_template(template="{input}.\nCurrent user data: {participant_data}")
+        chain = (
+            {"input": RunnablePassthrough()}
+            | RunnablePassthrough.assign(participant_data=RunnableLambda(lambda x: self.get_participant_data(state)))
+            | prompt
+            | super().get_chat_model().with_structured_output(json_schema)
+        )
+        output = chain.invoke(state["messages"][-1])
+        return output
+
+    def get_participant_data(self, state: PipelineState):
+        session = self.experiment_session(state)
+        return session.get_participant_data()
+
+    @staticmethod
+    def to_json_schema(data: dict):
+        """Converts a dictionary to a JSON schema by first converting it to a Pydantic object and dumping it again.
+        The input should be in the format {"key": "description", "key2": [{"key": "description"}]}
+
+        Nested objects are not supported at the moment
+
+        Input example 1:
+        {"name": "the user's name", "surname": "the user's surname"}
+
+        Input example 2:
+        {"name": "the user's name", "pets": [{"name": "the pet's name": "type": "the type of animal"}]}
+
+        """
+
+        def _create_model_from_data(value_data, model_name: str):
+            pydantic_schema = {}
+            for key, value in value_data.items():
+                if isinstance(value, str):
+                    pydantic_schema[key] = (str | None, Field(description=value))
+                elif isinstance(value, list):
+                    model = _create_model_from_data(value[0], key.capitalize())
+                    pydantic_schema[key] = (list[model], Field(description=f"A list of {key}"))
+            return create_model(model_name, **pydantic_schema)
+
+        Model = _create_model_from_data(data, "CustomModel")
+        schema = Model.model_json_schema()
+        # The schema needs a description in order to comply with function calling APIs
+        schema["description"] = ""
+        return schema
+
+
+class UpdateParticipantMemory(PipelineNode):
+    """A simple component to merge the the input data into the participant's memory data. If key_name is specified
+    the input data will be merged with `key_name` as the key name.
+    """
+
+    __human_name__ = "Update participant memory"
+    key_name: str | None = None
+
+    def _process(self, state: PipelineState) -> PipelineState:
+        extracted_data = state["messages"][-1]
+
+        if not isinstance(extracted_data, dict):
```

Comment:
If `key_name` is specified then it could be OK to get a string here and just set that value in the data.